### PR TITLE
fix(rnd): Added updated graph meta to include runs

### PR DESCRIPTION
**Pull request by Swiftyos**
### Background

Fixes the large number of requests on the monitoring page, by expanding the get graphs call to include minimal runs meta. Thus we get all the data for the monitoring page in a single request.

Also:
- Tidied up pyproject.toml file to remove none existent build steps
- Made sentry toggle able

### CODE REVIEW

File: rnd/autogpt_server/autogpt_server/data/graph.py [hunk 1]
```diff
@@ -77,6 +80,25 @@ def from_db(node: AgentNode):
...
 
 
+class ExecutionMeta(BaseDbModel):
+    execution_id: str
+    started_at: datetime
+    ended_at: datetime
+    status: autogpt_server.data.execution.ExecutionStatus
+
+    @staticmethod
+    def from_db(execution: AgentGraphExecution):
```
Comment (kcze) [inline rnd/autogpt_server/autogpt_server/data/graph.py]: This isn't from db, it's just a subset of data from `AgentGraphExecution`.

File: rnd/autogpt_server/autogpt_server/data/graph.py [hunk 2]
```diff
@@ -85,15 +107,26 @@ class GraphMeta(BaseDbModel):
     name: str
     description: str
 
+    executions: list[ExecutionMeta] = []
```
Comment (kcze) [inline rnd/autogpt_server/autogpt_server/data/graph.py]: nit: Redundant empty line (same above `name` but that isn't this PR)
```suggestion
    description: str
    executions: list[ExecutionMeta] = []
```

File: rnd/autogpt_builder/src/app/page.tsx [hunk 3]
```diff
@@ -17,62 +18,53 @@ import {
...
-      setFlows(flows);
-      flows.map((flow) => refreshFlowRuns(flow.id));
+      setAgentRuns(flowRuns);
     });
-  }, [api, refreshFlowRuns]);
+  }, [api]);
+
+  useEffect(() => {
+    fetchAgents();
+  });
```
Comment (kcze) [inline rnd/autogpt_builder/src/app/page.tsx]: Is this intended? This `useEffect` will run on every render. If you want to run it once:
```suggestion
  useEffect(() => {
    fetchAgents();
  }, []);
```

**Review by kcze — CHANGES_REQUESTED** [review]
This PR renames some things from `flow` to `agent` but there are still inconsistencies. I think using different words for the same thing is confusing for us and contributors.

File: rnd/autogpt_builder/src/app/page.tsx (hunk 3 above)
Comment (Swiftyos) [reply rnd/autogpt_builder/src/app/page.tsx]: No it was intended to run once on page load

File: rnd/autogpt_server/autogpt_server/data/graph.py (hunk 1 above)
Comment (Swiftyos) [reply rnd/autogpt_server/autogpt_server/data/graph.py]: I can rename that function 👍
That said the all functions with this name are inaccurate then as prisma already returns an object

**Comment by Swiftyos** [comment]
> > This PR renames some things from `flow` to `agent` but there are still inconsistencies. I think using different words for the same thing is confusing for us and contributors.
> 
> I agree, I was working on bring down the scope of concepts we have. We use Agent, Flow and Graph all interchangeable atm
> 
> Given we only talk to endusers with the term agent I was trying to standardise on that.
> 
> Would you like me to propagate the name change across the system?

I decided it was best to change the naming schema in a different PR

File: rnd/autogpt_builder/src/app/page.tsx (hunk 3 above)
Comment (Swiftyos) [reply rnd/autogpt_builder/src/app/page.tsx]: I think I've fixed it now?

File: rnd/autogpt_builder/src/app/page.tsx [hunk 4]
```diff
@@ -123,56 +97,30 @@ const Monitor = () => {
...
-    startTime,
-    endTime,
-    duration,
-    totalRunTime,
-    nodeExecutionResults: nodeExecutionResults,
-  };
+    startTime: new Date(executionMeta.started_at).getTime(),
+    endTime: executionMeta.ended_at
+      ? new Date(executionMeta.ended_at).getTime()
+      : undefined,
```
Comment (Pwuts) [inline rnd/autogpt_builder/src/app/page.tsx]: Parsing date strings from the API into `Date` should happen in the API client. It already does this in `getGraphExecutionInfo` and `stopGraphExecution`.

File: rnd/autogpt_builder/src/lib/autogpt-server-api/baseClient.ts [hunk 5]
```diff
@@ -44,8 +44,8 @@ export default class BaseAutoGPTServerAPI {
     return await this._get("/blocks");
   }
 
-  async listGraphs(): Promise<GraphMeta[]> {
-    return this._get("/graphs");
+  async listGraphs(with_runs: boolean = false): Promise<GraphMeta[]> {
+    return this._get(`/graphs?with_runs=${with_runs}`);
```
Comment (Pwuts) [inline rnd/autogpt_builder/src/lib/autogpt-server-api/baseClient.ts]: If it returns a different type depending on the parameter, pls differentiate the output type and add an overload.

File: rnd/autogpt_server/autogpt_server/data/graph.py [hunk 6]
```diff
@@ -77,23 +80,67 @@ def from_db(node: AgentNode):
...
+
+
 class GraphMeta(BaseDbModel):
     version: int = 1
     is_active: bool = True
     is_template: bool = False
-
     name: str
     description: str
+    executions: list[ExecutionMeta] = []
```
Comment (Pwuts) [inline rnd/autogpt_server/autogpt_server/data/graph.py]:
```suggestion
    executions: Optional[list[ExecutionMeta]] = None
```
To distinguish between a graph not having any executions vs the executions not being queried.

File: rnd/autogpt_server/autogpt_server/data/graph.py [hunk 7]
```diff
@@ -338,6 +385,7 @@ async def get_node(node_id: str) -> Node:
 
 
 async def get_graphs_meta(
+    include_executions: bool = False,
```
Comment (Pwuts) [inline rnd/autogpt_server/autogpt_server/data/graph.py]: Pls also update docstring

File: rnd/autogpt_server/autogpt_server/data/graph.py [hunk 8]
```diff
@@ -77,23 +80,67 @@ def from_db(node: AgentNode):
...
+        )
+
+
 class GraphMeta(BaseDbModel):
     version: int = 1
     is_active: bool = True
     is_template: bool = False
-
     name: str
     description: str
```
Comment (ntindle) [inline rnd/autogpt_server/autogpt_server/data/graph.py:123]: Don’t put arrays as defaults for functions it’s got wild side effects

File: rnd/autogpt_server/autogpt_server/data/graph.py [hunk 9]
```diff
@@ -361,11 +409,19 @@ async def get_graphs_meta(
...
+            distinct=["id"],
+            order={"version": "desc"},
+            include=AgentGraphInclude(AgentGraphExecution={"include": {"AgentNodeExecutions": True}}),  # type: ignore
+        )
+    else:
+        graphs = await AgentGraph.prisma().find_many(
+            where=where_clause,
+            distinct=["id"],
+            order={"version": "desc"},
+        )
```
Comment (Pwuts) [inline rnd/autogpt_server/autogpt_server/data/graph.py]: Unnecessary duplication and `# type: ignore`:
```suggestion
    graphs = await AgentGraph.prisma().find_many(
        where=where_clause,
        distinct=["id"],
        order={"version": "desc"},
        include=AgentGraphInclude(
            AgentGraphExecution={"include": {"AgentNodeExecutions": True}}
        )
        if include_executions
        else None,
    )
```

File: rnd/autogpt_builder/src/lib/autogpt-server-api/baseClient.ts (hunk 5 above)
Comment (Swiftyos) [reply rnd/autogpt_builder/src/lib/autogpt-server-api/baseClient.ts]: It does not return different type depending on the output.

File: rnd/autogpt_server/autogpt_server/data/graph.py (hunk 9 above)
Comment (Swiftyos) [reply rnd/autogpt_server/autogpt_server/data/graph.py]: type: ignore is cause the type system can't recognise a valid prisma query...

File: rnd/autogpt_server/autogpt_server/data/graph.py (hunk 6 above)
Comment (Swiftyos) [reply rnd/autogpt_server/autogpt_server/data/graph.py]: Makes sense okay, I'll add that

**Comment by majdyz** [comment]
I'm still seeing this:

https://github.com/user-attachments/assets/0574512a-9fd1-4e40-b374-3728f1ee3c99

Also seeing another error:


https://github.com/user-attachments/assets/dce4d799-b05f-4b70-a194-966133b43fbd

**Comment by majdyz** [comment]
oh nvm it's fixed on the latest changes

File: rnd/autogpt_builder/src/app/page.tsx (hunk 4 above)
Comment (Swiftyos) [reply rnd/autogpt_builder/src/app/page.tsx]: I've moved it into the api

File: rnd/autogpt_server/autogpt_server/data/graph.py (hunk 9 above)
Comment (Pwuts) [reply rnd/autogpt_server/autogpt_server/data/graph.py]: yes, it can:
image

File: rnd/autogpt_builder/src/lib/autogpt-server-api/types.ts [hunk 10]
```diff
@@ -158,6 +158,15 @@ export type GraphMeta = {
...
   description: string;
+};
+
+export type GraphMetaWithRuns = {
+  id: string;
+  version: number;
+  is_active: boolean;
+  is_template: boolean;
+  name: string;
+  description: string;
```
Comment (Pwuts) [inline rnd/autogpt_builder/src/lib/autogpt-server-api/types.ts]:
```suggestion
export type GraphMetaWithRuns = GraphMeta & {
```

File: rnd/autogpt_server/autogpt_server/data/graph.py [hunk 11]
```diff
@@ -409,19 +409,18 @@ async def get_graphs_meta(
...
+        where=where_clause,
+        distinct=["id"],
+        order={"version": "desc"},
+        include=(
+            AgentGraphInclude(
+                AgentGraphExecution={"include": {"AgentNodeExecutions": True}}
+            )
+            if include_executions
+            else None
+        ),  # type: ignore
```
Comment (Pwuts) [inline rnd/autogpt_server/autogpt_server/data/graph.py]: unnecessary `# type: ignore`

File: rnd/autogpt_server/pyproject.toml [hunk 12]
```diff
@@ -69,40 +70,6 @@ cli = "autogpt_server.cli:main"
...
-cmd = "python setup.py bdist_msi"
-
-[tool.poe.tasks.dist_appimage]
-cmd = "python setup.py bdist_appimage"
-
-[tool.poe.tasks.dist_deb]
-cmd = "python setup.py bdist_deb"
-
-[tool.poe.tasks._dinstall]
-cmd = "python setup.py install"
```
Comment (Pwuts) [inline rnd/autogpt_server/pyproject.toml]: Unrelated & undeclared change

File: rnd/autogpt_builder/src/lib/autogpt-server-api/baseClient.ts [hunk 13]
```diff
@@ -181,9 +188,17 @@ export default class BaseAutoGPTServerAPI {
           ? {
               "Content-Type": "application/json",
               Authorization: token ? `Bearer ${token}` : "",
+              "Access-Control-Allow-Origin": "*",
+              "Access-Control-Allow-Methods":
+                "GET, POST, PUT, PATCH, DELETE, OPTIONS",
+              "Access-Control-Allow-Headers": "Content-Type, Authorization",
```
Comment (Pwuts) [inline rnd/autogpt_builder/src/lib/autogpt-server-api/baseClient.ts]: These are response headers, not request headers. Same below.

File: rnd/autogpt_server/pyproject.toml (hunk 12 above)
Comment (Swiftyos) [reply rnd/autogpt_server/pyproject.toml]: I added it to the PR description

File: rnd/autogpt_server/autogpt_server/data/graph.py (hunk 11 above)
Comment (Swiftyos) [reply rnd/autogpt_server/autogpt_server/data/graph.py]: Fixed, yeah problem was with my broken env

**Review by Pwuts — DISMISSED** [review]
LGTM!

File: rnd/autogpt_builder/sentry.edge.config.ts [hunk 14]
```diff
@@ -8,6 +8,8 @@ import * as Sentry from "@sentry/nextjs";
 Sentry.init({
   dsn: "https://fe4e4aa4a283391808a5da396da20159@o4505260022104064.ingest.us.sentry.io/4507946746380288",
```
Comment (ntindle) [inline rnd/autogpt_builder/sentry.edge.config.ts]: I’d like to suggest doing the before send instead of the whole thing and see how you feel about that

Comment (Swiftyos) [reply rnd/autogpt_builder/sentry.edge.config.ts]: I don't understand what you are getting at?

Comment (ntindle) [reply rnd/autogpt_builder/sentry.edge.config.ts]: You can disable the popup instead of all of sentry if desired

**Comment by majdyz** [comment]
@Swiftyos This fix seems to be crucial.
 I have resolved the conflict on this PR, try to rebase and let me know if it is working as expected.

**Review by majdyz — APPROVED** [review]

**Comment by Swiftyos** [comment]
Tested these routes merging
- [X] Create from scratch and execute an agent with at least 3 blocks
- [X] Import an agent from file upload, and confirm it executes correctly
- [X] Upload agent to marketplace
- [X] Edit an agent from monitor, and confirm it executes correctly

**Pull request merged**